bip-0322.mediawiki

```diff
@@ -13,119 +13,99 @@
 
-=== Signing ===
+A validator must verify it is valid and meets the description of virtual transactions as specified above. See "Validation" below.
 
-The signature is generated as follows:
+=== Validation ===
 
-# Derive the private key privkey for the scriptPubKey; FAIL if not VALID
-# Generate and return a signature sig with privkey=privkey, sighash=sighash
-
-=== Verifying ===
+To validate a proof with additional inputs, the following steps must be taken:
 
-Verify a proof, given a standard flags value, a script sig, an optional witness, and a derived sighash as described above.
+# deserialize the to_spend and to_sign transactions from the proof, and fail if the proof contains extraneous bytes
+# verify that any additional inputs being proven (proof of funds) are included as inputs in the to_sign transaction, exactly once
+# reconstruct the to_spend' and to_sign' transactions, based on the specification above, copying the version, lock time, and sequence values
+# verify that to_spend = to_spend', that to_sign has at least 1 input, has exactly 1 output, and that to_sign.vin[0] = to_sign'.vin[0]
+# optional: set the "in_future" flag if the transaction's lock time is in the future according to consensus rules
+# in "coins", a mapping of outpoints (hash, vout) to coins (scriptPubKey, amount), let coins(to_spend.txid, 0) = (to_spend.vout[0], 0)
+# for each proof of fund input, set the corresponding values in the coins map; abort if the input cannot be found
+# check the signature of each input using consensus rules, then upgradable rules
 
-While omitted below, ERROR is returned if an unforeseen error occurs at any point in the process. A concrete example of this is if a legacy proof is given as input to a non-legacy address; the deserialization of the proof will fail in this case, and this should result in an ERROR result.
+To validate a simple proof, the following steps must be taken:
 
-# Verify Script with flags=consensus flags (currently P2SH, DERSIG, NULLDUMMY, CLTV, CSV, WITNESS), scriptSig=script sig, scriptPubKey=scriptPubKey, witness=witness, and sighash=sighash
-# Return INVALID if verification fails
-# Verify Script with flags=standard flags (above plus STRICTENC, MINIMALDATA, etc.), scriptSig=script sig, scriptPubKey=scriptPubKey, witness=witness, and sighash=sighash
-# Return VALID if verification succeeds, otherwise return INCONCLUSIVE
+# construct the to_spend and to_sign transactions, based on the specification above
+# check the signature using consensus rules, then upgradable rules
 
 == Legacy format ==
 
-The legacy format is restricted to the legacy P2PKH address format.
+The legacy format is restricted to the legacy P2PKH invoice address format.
 
-Any other input (i.e. non-P2PKH address format) must be signed using the new format described above.
+Any other input (i.e. non-P2PKH invoice address format) must be signed using the new format described above.
```

Comment:
Perhaps it should be recommended here to prefer the general format even for P2PKH going forth.

Reply:
This was discussed in an earlier draft of this proposal, and then people were opposing the conclusion (that was to use legacy for P2PK(H)). I'm going to push for using the new format and see what people think, so:

> New proofs should use the new format for all invoice address formats, including P2PKH.
>
> The legacy format MAY be used, but must be restricted to the legacy P2PKH invoice address format.